pm/recipient.go

```diff
@@ -0,0 +1,223 @@
+		return false, err
+	}
+
+	if !r.validRand(recipientRand) {
+		return false, errors.Errorf("invalid already revealed recipientRand %v", recipientRand)
+	}
+
+	if err := r.updateSenderNonce(recipientRand, ticket.SenderNonce); err != nil {
+		return false, err
+	}
+
+	if r.val.IsWinningTicket(ticket, sig, recipientRand) {
+		if err := r.store.Store(ticket.RecipientRandHash.Hex(), ticket, sig, recipientRand); err != nil {
+			return true, err
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// RedeemWinningTicket redeems all winning tickets with the broker
+// for a session ID
+func (r *recipient) RedeemWinningTickets(sessionID string) error {
+	tickets, sigs, recipientRands, err := r.store.Load(sessionID)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(tickets); i++ {
+		if err := r.redeemWinningTicket(tickets[i], sigs[i], recipientRands[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TicketParams returns the recipient's currently accepted ticket parameters
+func (r *recipient) TicketParams(sender ethcommon.Address) (*TicketParams, error) {
+	randBytes := make([]byte, 32)
+	if _, err := rand.Read(randBytes); err != nil {
+		return nil, err
+	}
+
+	seed := new(big.Int).SetBytes(randBytes)
+	recipientRand := r.rand(seed, sender)
+	recipientRandHash := crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size))
+
+	return &TicketParams{
+		FaceValue:         r.faceValue,
+		WinProb:           r.winProb,
+		RecipientRandHash: recipientRandHash,
+		Seed:              seed,
+	}, nil
+}
+
+func (r *recipient) redeemWinningTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) error {
+	deposit, err := r.broker.GetDeposit(ticket.Sender)
+	if err != nil {
+		return err
+	}
+
+	penaltyEscrow, err := r.broker.GetPenaltyEscrow(ticket.Sender)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Consider a smarter strategy here in the future
+	// Ex. If deposit < transaction cost, do not try to redeem
+	if deposit.Cmp(big.NewInt(0)) == 0 && penaltyEscrow.Cmp(big.NewInt(0)) == 0 {
+		return errors.Errorf("sender %v has zero deposit and penalty escrow", ticket.Sender)
+	}
+
+	// Assume that that this call will return immediately if there
+	// is an error in transaction submission. Else, the function will kick off
+	// a goroutine and then return to the caller
+	if err := r.broker.RedeemWinningTicket(ticket, sig, recipientRand); err != nil {
+		return err
+	}
+
+	// If there is no error, the transaction has been submitted. As a result,
+	// we assume that recipientRand has been revealed so we should invalidate it locally
+	r.updateInvalidRands(recipientRand)
+
+	// After we invalidate recipientRand we can clear the memory used to track
+	// its latest senderNonce
+	r.clearSenderNonce(recipientRand)
+
+	return nil
+}
+
+func (r *recipient) rand(seed *big.Int, sender ethcommon.Address) *big.Int {
+	h := hmac.New(sha256.New, r.secret[:])
+	h.Write(append(seed.Bytes(), sender.Bytes()...))
```

Comment:
Question of ignorance :) what does the `...` at the end of the `append` call do?

Reply:
`sender.Bytes()...` destructures `[]byte` from a slice into a list of arguments i.e. individual bytes which is necessary or else `append` will add `sender.Bytes()` as a single element to the slice `seed.Bytes()` when we really want `append` to add each byte of `sender.Bytes()` as its own element.